diesel/src/r2d2.rs

```diff
@@ -175,6 +192,13 @@ where
     }
 }
 
+/// Trait to provide the user the option to change parameters of the URL at runtime.
+/// E.g. to implement password rotation
+pub trait UrlProvider: Send + Sync + fmt::Debug + 'static {
```

Comment:
Why just do not use `Fn/FnOnce/FnMut` trait?

Reply:
For better customizability of the provider.

Reply:
For example a provider such as this one would enable password rotation.

```rust
#[derive(Clone, Debug)]
struct CustomUrlProvider {
    password: Arc<RwLock<String>>,
    host: String,
    port: u32,
    user: String,
    database: String,
    region: Region
}

impl CustomUrlProvider {

    // ...

    fn rotate_password(&self) -> Result<()> {
        let new_token = crate::iam_authentication::generate_db_auth_token(
            &self.region,
            &self.host,
            &self.port,
            &self.user,
            &WebIdentityProvider::from_k8s_env()
        );

        if let Ok(mut pw) = self.password.write() {
            *pw = crate::iam_authentication::urlencode(&new_token?);
            Ok(())
        } else {
            Err(CustomError::DbPasswordRotationError())
        }
    }
}

impl UrlProvider for CustomUrlProvider {
    fn provide_url(&self) -> String {
        loop {
            if let Ok(pw) = self.password.read() {
                break format!(
                    "postgres://{}:{}@{}:{}/{}",
                    self.user, *pw, self.host, self.port, self.database
                );
            } else {
                warn!("Couldn't acquire lock to build DB password")
            }
        }
    }
}
```

Reply:
I think it should be easy to just provide a default implementation all types that implement `Fn()` somehow.

Reply:
Sure, it's possible to use a closure instead. Nonetheless, to me it seems more handy to use a new trait. A custom provider can have a state and easily be dealing with connection updates or different test scenarios. 

On the contrary, it's really just about making the db connection more dynamic and a `Fn()` would suffice.

Reply:
To be clear: I'm not talking about removing `UrlProvider`, just about adding a generic `impl<T:Fn()> UrlProvider for T` to allow both variants.